Altered rwf_io to be able to write directly in root

andLaing opened the pull request:
EArray for the sensors can now be saved
in h5file.root as is done in the C++
detsim.

Tidied and added type hints, docstrings
and a test for the writer.

andLaing committed:
Altered rwf_io to be able to write directly in root

EArray for the sensors can now be saved
in h5file.root as is done in the C++
detsim.

Tidied and added type hints, docstrings
and a test for the writer.

andLaing committed:
removed some whitespace

gonzaponte commented on invisible_cities/io/rwf_io.py:
```diff
@@ -1,23 +1,57 @@
-    except tb.NoSuchNodeError: rwf_group = file.create_group(file.root, group_name)
+               group_name      :          str          ,
+               table_name      :          str          ,
+               compression     :          str = 'ZLIB4',
+               n_sensors       :          int          ,
+               waveform_length :          int          ) -> Callable:
```
This will not work, arguments with default values (optional) cannot be followed by arguments without default values (mandatory)

andLaing replied:
Curiously, that's not something that I changed but now that you mention it it shouldn't work. Must just be that we've always set everything explicitly or something. I'll change it to default to the NEW numbers

gonzaponte replied:
You are right, it was already like that. I've only noticed it because you changed something else :laughing: 
If you want, leave it as is and we open a new PR fixing it everywhere. I don't think it should take any particular number.

andLaing replied:
OK, I'll leave it for now

gonzaponte commented on invisible_cities/io/rwf_io.py:
```diff
@@ -1,23 +1,57 @@
+                      Number of samples per sensor
+    """
+    if group_name is 'root':
+        rwf_group = h5out.root
```
I think it is more natural to use an empty string than `"root"`, but it is less comprehensible. Maybe using `None` is a good compromise.

gonzaponte commented on invisible_cities/io/rwf_io.py:
```diff
@@ -1,23 +1,57 @@
+        try:
+            rwf_group = getattr           (h5out.root, group_name)
+        except tb.NoSuchNodeError:
+            rwf_group = h5out.create_group(h5out.root, group_name)
```
I think it is more correct to ask whether the node exists than trying to see if it works:
```
if group_name in h5out.root:
    h5out.create_group(h5out.root, group_name)

rwf_group = getattr(h5out.root, group_name)
```

gonzaponte commented on invisible_cities/io/rwf_io.py:
```diff
@@ -1,23 +1,57 @@
                                    table_name,
                                    atom    = tb.Int16Atom(),
                                    shape   = (0, n_sensors, waveform_length),
                                    filters = tbl.filters(compression))
```
slight missalignment

gonzaponte commented on invisible_cities/io/rwf_io_test.py:
```diff
@@ -0,0 +1,45 @@
+            rwf_writer_(sens)
+
+    with tb.open_file(ofile) as h5test:
+        if group_name is 'root':
```
It is not correct to compare strings with `is`. `is` checks whether two objects have the same ID, i.e. if they refer to the same memory space. Curiously this works, interactively, but it is not guaranteed to work in general.

jacg replied:
> Curiously this works, interactively

Yes, it depends on whether the string is interned or not ... and that depends on complex implementation-defined details (which are probably not guaranteed not to evolve with time).  So, to a first order approximation:

1. Do **not** use `is` to compare strings ...
2. ... unless 
    + you have benchmarked to prove that your code is too slow
    + you have profiled to prove that some string comparison is taking a significant portion of your CPU cycles

    If you ever find yourself in that situation (rather unlikely) then interning relevant strings and using `is` might speed it up.

TL;DR: do not use `is` to compare strigns.

gonzaponte commented on invisible_cities/io/rwf_io_test.py:
```diff
@@ -0,0 +1,45 @@
+
+        table = getattr(group, table_name)
+        assert table.shape == (nevt, nsensor, nsample)
+        assert_allclose(test_data, table.read())
```
since they are ints, you should be able to compare them exactly, right?

gonzaponte reviewed (commented):
I can't believe we didn't have `rwf_io_test.py`!

andLaing committed:
Changed rwf_io group definition to if else and root group to None

Changed the try except structure to instead
check if the group already exists.

Changed 'root' for None in root group definition.

Some minor cosmetics.

andLaing committed:
Removed string comparison with is and added equality comparison

test_rwf_writer changedto remove potentially
unstable 'is' comparison of strings.

Outputs now compared with np.all(arr1 == arr2)

gonzaponte reviewed (approved):
Nice extension to the rwf writer. Now it allows writing to the root group. It also adds a good test that covers both the new and the original cases, which were somehow untested.
Good job!
